File: java/src/main/java/br/com/emendes/problems/RemoveDuplicatesFromSortedArrayII.java
package br.com.emendes.problems;

/**
 * Dado um array de inteiros nums ordenado em ordem crescente,
 * remova algumas duplicatas in-place, de modo que cada elemento exclusivo apareça no máximo duas vezes.
 * A ordem relativa dos elementos deve ser mantida a mesma.<br><br>
 * <p>
 * Como é impossível alterar o comprimento do array em algumas linguagens,
 * você deve colocar o resultado na primeira parte do array nums.
 * Mais formalmente, se houver k elementos após a remoção das duplicatas,
 * então os primeiros k elementos de nums deverão conter o resultado final.
 * Não importa o que você deixa além dos primeiros k elementos.<br><br>
 * <p>
 * Retorne k depois de colocar o resultado final nos primeiros k espaços de nums.<br><br>
 * <p>
 * Não aloque espaço extra para outro array.
 * Você deve fazer isso modificando o array de entrada local com memória extra O(1).<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= nums.length <= 3 * 10⁴<br>
 * -10⁴ <= nums[i] <= 10⁴<br>
 * nums está ordenado em ordem não decrescente.
 */
public class RemoveDuplicatesFromSortedArrayII {

  public int removeDuplicates(int[] nums) {
    int counter = 1;
    int k = 1;
    for (int i = 1; i < nums.length; i++) {
      if (nums[i] == nums[i - 1]) {
        counter++;
      } else {
        counter = 1;
      }
      if (counter <= 2) {
        nums[k] = nums[i];
        k++;
      }
    }

    return k;
  }

}

File: javascript/problems/RemoveDuplicatesFromSortedArrayII.js
/**
 * Dado um array de inteiros nums ordenado em ordem crescente, remova algumas duplicatas in-place
 * de modo que cada único elemento apareça no máximo 2 vezes. A ordem relativa dos elementos deve ser mantida
 * a mesma.
 *
 * Já que é impossível mudar o comprimento de um array em algumas linguagens, em vez disso,
 * você deve colocar o resultado na primeira parte do array nums.
 * Mais formalmente, se houver k elementos após a remoção das duplicatas,
 * os primeiros k elementos de nums devem conter o resultado final.
 * Não importa o que você deixe além dos primeiros k elementos.
 *
 * Retorna k após colocar o resultado final nos primeiros k slots de nums.
 *
 * Não aloque espaço extra para outro array. Você deve fazer isso modificando a matriz
 * de entrada no local com O(1) de memória extra.
 *
 * Restrições:
 *
 * 1 <= nums.length <= 3 * 10⁴
 *
 * -10⁴ <= nums[i] <= 10⁴
 *
 * nums está ordenado em ordem crescente.
 */
const removeDuplicates = function (nums) {
  let prev = nums[0];
  let k = 1;
  let count = 1;
  for (let i = 1; i < nums.length; i++) {
    if (nums[i] === prev) {
      count++;
      if (count < 3) nums[k++] = nums[i];
    } else {
      prev = nums[i];
      count = 1;
      nums[k++] = nums[i];
    }
  }

  return k;
};
